Leave out .js so Metro can resolve native variant

hypest opened the pull request:
## Description
This PR drops the `.js` extension from `@wordpress/element`'s `package.json` entrypoint definitions so when used in the mobile RN app the correct module (`index.native.js`) can be resolved by Metro.

The same treatment should probably be performed in all packages, or at least the ones that have native mobile specializations but, let's do that step by step starting with the `element` package that currently is breaking the RN app (at the time of writing, [commit hash](d73c4431f44430b345ab18c4622d916ad5d78dd4)).

## How has this been tested?
Ran `npm run dev` successfully.

## Types of changes
Bug fix: Modify the `@wordpress/element` package entrypoint definitions (`main`, `module`).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

hypest committed:
Leave out .js so Metro can resolve native variant

gziolo commented:
We can also try using react-native field instead as suggested here: https://github.com/facebook/metro/issues/59#issuecomment-330530676.

For some reasons they usually add an extension in the path provided in the main field. I’m not sure why. We probably should  to contact npm support to confirm it.

hypest commented:
Thanks for the research @gziolo ! I hated to have the `.js` removed but that was the only solution I came up with after failing to find a canonical one :(

I will test out the `react-native` field option and will revise this PR if successful.

hypest committed:
Revert "Leave out .js so Metro can resolve native variant"

This reverts commit e2de6d208b1bc47c698690e1c2d730d092148139.

hypest committed:
Use the `react-native` entrypoint for RN

And let it point to the extension-less index module so Metro can decide
which module to resolve to, as normal.

hypest commented:
Reverted the previous change and implemented the solution based on the `react-native` package.json entry point @gziolo , ready for a review!

I tested the solution against [this PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/61) over at the RN app repo.

gziolo reviewed (commented):
Technically, you can use `build/index.native.js` as we don't have other overrides. It's colocated with the code, so you know upfront what options are there. If you prefer it, we can leave it as is 👍 

Do we need to add `react-native` fields to other packages that use RN overrides? I'd prefer to include them all in this PR.

hypest commented:
> Technically, you can use build/index.native.js as we don't have other overrides. It's colocated with the code, so you know upfront what options are there. If you prefer it, we can leave it as is 👍

True, we can also point to the full filename. I'm not yet 100% which approach is best for the RN case; we should keep an open eye on this. Let's leave it extension-free for now since it feels more Metro-native (but not as much npm-native) and we can revise in the future if new info comes up.

> Do we need to add react-native fields to other packages that use RN overrides? I'd prefer to include them all in this PR.

Currently, there is no other package in the `packages` folder that has native mobile files. That said, we can still add the entrypoint anyway (provided we use the extension-free form) and will work just fine since Metro will pick up the web variant as normal. Let me know if you want to go that route.

hypest commented:
@gziolo , I just found out that the `react-native` field/entrypoint is not what's picked up by the jest tests of the RN app repo. Also tried the `browser` entrypoint but doesn't work either. One needs to leave the `main` entrypoint extension-less for the jest tests to succeed. So, we need some more thinking to do on this one.

gziolo commented:
I see, I wrote to one of the maintainers of Jest, let's see if they have any solution ready. I couldn't find anything like that in their repository.

hypest commented:
> I just found out that the react-native field/entrypoint is not what's picked up by the jest tests of the RN app repo.

Another false negative @gziolo 😢 . I've updated the RN app (see https://github.com/wordpress-mobile/gutenberg-mobile/pull/61#issuecomment-402895217) to consume packages directly from `/packages` and to point to the latest from #7691 and both the RN app and the testsuite is green. I think we can probably lower the priority of fixing this.

We'll still want to remove the highly undocumented `react-native` entry point and use a more robust solution based on the `main` entrypoint, but we leave it aside for a bit.
